### PR TITLE
Possible fix for #254 - be tolerant of nan chunks.

### PR DESCRIPTION
**Pull request by JSKenyon**
- [X] Tests added / passed

  ```bash
  $ py.test -v -s daskms/tests
  ```

  If the pep8 tests fail, the quickest way to correct
  this is to run `autopep8` and then `flake8` and
  `pycodestyle` to fix the remaining issues.

  ```
  $ pip install -U autopep8 flake8 pycodestyle
  $ autopep8 -r -i daskms
  $ flake8 daskms
  $ pycodestyle daskms
  ```

- [x] Fully documented, including `HISTORY.rst` for all changes
      and one of the `docs/*-api.rst` files for new API

  To build the docs locally:

  ```
  pip install -r requirements.readthedocs.txt
  cd docs
  READTHEDOCS=True make html
  ```

### CODE REVIEW

**Comment by JSKenyon** [comment]
@bennahugo Could you try this out? I don't love this solution but couldn't think of a better one (I hate if-else ladders).  

@sjperkins can you please take a look when you feel inclined? Maybe you have a better solution. All tests failing on minio install at present.

**Comment by bennahugo** [comment]
sure I can. Give me a minute

**Comment by sjperkins** [comment]
Can you rebase on master? A fix for the CI breakage has been merged.

**Comment by bennahugo** [comment]
does not look like this fixes the issue yet
```
table = '/tmp/tmpgdapi_6q_averaged.ms'
table_proxy = TableProxy[/tmp/tmpgdapi_6q_averaged.ms](table, /tmp/tmpgdapi_6q_averaged.ms, ack=False,readonly=False,lockoptions=user,__executor_key__=/tmp/tmpgdapi_6q_averaged.ms)
datasets = [], columns = ['ANTENNA1', 'ANTENNA2', 'ARRAY_ID', 'DATA', 'DATA_DESC_ID', 'EXPOSURE', ...], descriptor = 'ms(False)', table_keywords = None
column_keywords = None

    def _write_datasets(
        table, table_proxy, datasets, columns, descriptor, table_keywords, column_keywords
    ):
        _, table_name, subtable = table_path_split(table)
        table_name = "::".join((table_name, subtable)) if subtable else table_name
        row_orders = []
    
        # Put table and column keywords
        table_proxy.submit(
            _put_keywords, WRITELOCK, table_keywords, column_keywords
        ).result()
    
        # Sort datasets on (not has "ROWID", index) such that
        # datasets with ROWID's are handled first, while
        # those without (which imply appends to the MS)
        # are handled last
        sorted_datasets = sorted(
            enumerate(datasets), key=lambda t: ("ROWID" not in t[1].data_vars, t[0])
        )
    
        # Establish row orders for each dataset
        for di, ds in sorted_datasets:
            try:
                rowid = ds.ROWID.data
            except AttributeError:
                # Add operation
                # No ROWID's, assume they're missing from the table
                # and remaining datasets. Generate addrows
                # NOTE(sjperkins)
                # This could be somewhat brittle, but exists to
                # update MS empty subtables once they've been
                # created along with the main MS by a call to default_ms.
                # Users could also it to append rows to an existing table.
                # An xds_append_to_table may be a better solution...
                last_datasets = datasets[di:]
                last_row_orders = add_row_order_factory(table_proxy, last_datasets)
    
                # We don't inline the row ordering if it is derived
                # from the row sizes of provided arrays.
                # The range of possible dependencies are far too large to inline
                row_orders.extend([(False, lro) for lro in last_row_orders])
                # We have established row orders for all datasets
                # at this point, quit the loop
                break
            else:
                # Update operation
                # Generate row orderings from existing row IDs
                row_order = cached_row_order(rowid)
    
                # Inline the row ordering in the graph
                row_orders.append((True, row_order))
    
        assert len(row_orders) == len(datasets)
    
        datasets = []
    
        for (di, ds), (inline, row_order) in zip(sorted_datasets, row_orders):
            # Hold the variables representing array writes
            write_vars = {}
    
            # Generate a dask array for each column
            for column in columns:
                try:
                    variable = ds.data_vars[column]
                except KeyError:
                    log.warning("Ignoring '%s' not present " "on dataset %d" % (column, di))
                    continue
                else:
                    full_dims = variable.dims
                    array = variable.data
    
                if not isinstance(array, da.Array):
                    raise TypeError(
                        "%s on dataset %d is not a dask Array "
                        "but a %s" % (column, di, type(array))
                    )
    
                args = [row_order, ("row",)]
    
                # We only need to pass in dimension extent arrays if
                # there is more than one chunk in any of the non-row columns.
                # In that case, we can putcol, otherwise putcolslice is required
    
                inlinable_arrays = [row_order]
    
                if (
                    row_order.shape[0] != array.shape[0]
                    or row_order.chunks[0] != array.chunks[0]
                ):
>                   raise ValueError(
                        f"ROWID shape and/or chunking does " f"not match that of {column}"
                    )
E                   ValueError: ROWID shape and/or chunking does not match that of ANTENNA1

../dask-ms/daskms/writes.py:651: ValueError

```

**Comment by bennahugo** [comment]
wait being stupid. need to force a install

**Comment by sjperkins** [comment]
> does not look like this fixes the issue yet

It doesn't look like that Exception is failing on the changes.  Could you check that you have the correct branch again?

**Comment by bennahugo** [comment]
yup this fixes it
```
2022-10-04 10:07:35 | INFO     | subtables:copy_subtables | Copying /tmp/tmptth8zoue.ms/STATE to /tmp/tmptth8zoue_averaged.ms/STATE
[########################################] | 100% Completed | 1.50 sms
2022-10-04 10:07:37 | INFO     | app:_execute_graph | Averaging Complete
2022-10-04 10:07:37 | CRITICAL | fixvis:fixms | UVW recompute does not support time-variable reference centres. Your dataset will contain averaged UVW coordinates!
PASSED
xova/apps/xova/tests/test_argparsing.py::flake-8::FLAKE8 SKIPPED (file(s) previously passed FLAKE8 checks)
xova/apps/xova/tests/test_argparsing.py::test_parse_fields[1, 3C147, 2-result0] PASSED
xova/apps/xova/tests/test_argparsing.py::test_parse_fields[PKS-1934-result1] PASSED

```

**Comment by bennahugo** [comment]
thanks a lot

File: daskms/writes.py [hunk 1]
```diff
@@ -644,13 +644,24 @@ def _write_datasets(
...
+                    )
+            else:
+                if (
+                    row_order.shape[0] != array.shape[0]
+                    or row_order.chunks[0] != array.chunks[0]
+                ):
+                    raise ValueError(
+                        f"ROWID shape and/or chunking does "
+                        f"not match that of {column}"
+                    )
```
Comment (sjperkins) [inline daskms/writes.py:664]: How about?

```python
if (np.isnan(row_order.shape[0]) != np.isnan(array.shape[0]) or
   row_order.shape[0] != array.shape[0] or
   any(np.isnan(a) != np.isnan(b)  or a != b for a, b in zip(row_order.chunks[0],array.chunks[0]))):

      raise ValueError(...)
```

Comment (JSKenyon) [reply daskms/writes.py:664]: Hmmm, more concise but quite confusing. Will think about it for a moment.

Comment (JSKenyon) [reply daskms/writes.py:664]: I like this compact expression but getting it to do the correct thing in call cases is difficult.

Comment (sjperkins) [reply daskms/writes.py:664]: > This doesn't work as written, as the case second conditional will always evaluate to True for nan chunks.

Good catch. How about this?

```python
if not (np.isnan(row_order.shape[0]) if np.isnan(array.shape[0]) else row_order.shape[0] == array.shape[0] and
   all(np.isnan(a) if np.isnan(b)  else a == b for a, b in zip(row_order.chunks[0],array.chunks[0]))):

      raise ValueError(...)

```

Comment (JSKenyon) [reply daskms/writes.py:664]: I think it works but I have no idea how to format it. Hahaha.

Comment (JSKenyon) [reply daskms/writes.py]: Ok, applied this change. If you are happy, go ahead and merge.

Comment (sjperkins) [reply daskms/writes.py]: Yeah apologies, I'm now using a pre-commit hook everywhere.  For future reference

```
(venv) $ pip install pre-commit
(venv) $ pre-commit install
(venv) $ pre-commit run --all-files
```

I should probably put this in the docs somewhere

**Comment by sjperkins** [comment]
Thanks for the fix @JSKenyon

**Pull request merged**